frontend/app/controllers/spree/checkout_controller.rb

```diff
@@ -161,6 +162,19 @@ def ensure_sufficient_stock_lines
       end
     end
 
+    def ensure_sufficient_shipment_stock_quantities
+      if @order.passed_checkout_step?('delivery')
```

Comment:
@kennyadsl It's not necessarily assuming that the following step is the last... but it must be executed after the order passed the `delivery` step, for some reasons.  

I'll share some context regarding why I added this `if` clause, I think it may help us in finding a better solution for the issue you exposed. 

The reason is that, without that `if`, this [spec](https://github.com/solidusio/solidus/blob/master/frontend/spec/features/checkout_unshippable_spec.rb) fails. The spec seems to be still relevant to me, so adding the `if` allowed me to add the functionality while still keeping the existing behavior.  

Today I tried to dig deeper in the issue regarding the spec mentioned above, and this is what I found out. After moving my `before_action` to the end of the chain things improved a little, but not enough. 

The reason for having that specific `if` is that the `before_action` `CheckoutController::before_payment` may manipulate the order, specifically these lines:

```ruby
        @differentiator = Spree::Stock::Differentiator.new(@order, packages)
        @differentiator.missing.each do |variant, quantity|
          @order.contents.remove(variant, quantity)
        end
```

Besides, the line `@order.contents.remove(variant, quantity)` has the side effect of changing the order state from `payment` to `address` 🤔. Have not checked why, but I guess some `ActiveRecord` callback is responsible for that. 

Missing contents must be removed before running the new `before_action`, or behavior will change, and as said before, the removal happens only when the customer tries to see the `checkout/payment` page, which happens after `delivery` is completed. I think the state ordering `delivery -> payment` is quite robust, as generally order total should be calculated only after shipping charges and taxes are known, both depending on addresses. But YMMV, so robust may not be enough.

Another option would be to move the new code in the `before_payment` method, so it's executed when it starts to be relevant, but that would limit its execution only to that precise step, eventually limiting its usefulness, that's why I think this is not a viable solution. 

The current solution has the strong disadvantage that `@order.passed_checkout_step?('delivery')` does not make clear why the step must be passed, and the strong connection with the existing `before_payment` method. 

A different approach may be to leverage the existing code in `before_payment` and change the `if` to this, for example:

```ruby
      packages = @order.shipments.map(&:to_package)
      @differentiator = Spree::Stock::Differentiator.new(@order, packages)
      if @differentiator.missing.empty?
      # ...
```

This way, there's no more explicit dependence on order steps and their names. What do you think?